File: MIDTERM-LAB-WORK-3/MP-18/removeemptyrows.js
/**
 * MP18 – Remove Rows with Empty Fields
 *
 * This Node.js program reads the CSV exam dataset, identifies rows that
 * contain one or more empty/blank fields, removes them, and writes the
 * cleaned dataset to a new output CSV file. A summary is displayed.
 *
 * Dataset: Exam results from University of Perpetual Help System - Molino
 * Columns: Candidate, Type, Exam, Language, Exam Date, Score, Result, Time Used
 *
 * Run: node MP18_RemoveEmptyRows.js
 */

const fs       = require('fs');       // File System module for reading/writing files
const path     = require('path');     // Path utilities
const readline = require('readline'); // For interactive console input

// ──────────────────────────────────────────────────────────────────
// parseCSVLine – Splits a CSV row respecting quoted fields
//   @param  {string}   line  – a single raw CSV line
//   @return {string[]}       – array of field strings
// ──────────────────────────────────────────────────────────────────
function parseCSVLine(line) {
    const tokens = [];
    let current  = '';
    let inQuotes = false; // true when inside a double-quoted field

    for (const ch of line) {
        if (ch === '"') {
            inQuotes = !inQuotes; // toggle quoted mode
        } else if (ch === ',' && !inQuotes) {
            tokens.push(current.trim());
            current = ''; // reset buffer for next field
        } else {
            current += ch;
        }
    }
    tokens.push(current.trim()); // push the last field
    return tokens;
}

// ──────────────────────────────────────────────────────────────────
// readCSV – Reads and parses the dataset CSV
//   @param  {string}   filePath – path to input CSV file
//   @return {Object[]}          – array of exam record objects
// ──────────────────────────────────────────────────────────────────
function readCSV(filePath) {
    const raw   = fs.readFileSync(filePath, 'utf8');  // read file content
    const lines = raw.split(/\r?\n/);                  // split into lines

    const records   = [];    // accumulate parsed records here
    let headerFound = false; // flag: have we passed the column header row?

    for (const line of lines) {
        if (!line.trim()) continue; // skip blank lines

        // The column-header row contains "Candidate" keyword
        if (!headerFound) {
            if (line.includes('Candidate')) headerFound = true;
            continue; // skip title/metadata rows
        }

        const tokens = parseCSVLine(line);
        if (tokens.length < 9) continue; // skip malformed rows

        // Map token positions to field names
        const candidate = tokens[0];
        const type      = tokens[1];
        const exam      = tokens[3];
        const language  = tokens[4];
        const examDate  = tokens[5];
        const score     = tokens[6];
        const result    = tokens[7];
        const timeUsed  = tokens[8] || '';

        // Include ALL rows (even those with empty fields) for detection
        records.push({ candidate, type, exam, language, examDate, score, result, timeUsed });
    }

    return records;
}

// ──────────────────────────────────────────────────────────────────
// getEmptyFields – Returns array of field names that are empty
//   @param  {Object}   record – an exam record object
//   @return {string[]}        – list of empty field names
// ──────────────────────────────────────────────────────────────────
function getEmptyFields(record) {
    const fields = [
        { name: 'Candidate', value: record.candidate },
        { name: 'Type',      value: record.type      },
        { name: 'Exam',      value: record.exam      },
        { name: 'Language',  value: record.language  },
        { name: 'Exam Date', value: record.examDate  },
        { name: 'Score',     value: record.score     },
        { name: 'Result',    value: record.result    },
        { name: 'Time Used', value: record.timeUsed  },
    ];
    // Filter to only the fields where value is empty/blank
    return fields.filter(f => !f.value).map(f => f.name);
}

// ──────────────────────────────────────────────────────────────────
// writeCleanCSV – Writes cleaned records to a new CSV file
//   @param {Object[]} records    – array of clean exam record objects
//   @param {string}   outputPath – destination file path
// ──────────────────────────────────────────────────────────────────
function writeCleanCSV(records, outputPath) {
    // Build CSV header
    const lines = ['Candidate,Type,Exam,Language,Exam Date,Score,Result,Time Used'];

    // Build each data row (wrap fields in quotes to handle commas)
    for (const r of records) {
        lines.push(
            `"${r.candidate}","${r.type}","${r.exam}","${r.language}",` +
            `"${r.examDate}","${r.score}","${r.result}","${r.timeUsed}"`
        );
    }

    // Write all lines to the output file
    fs.writeFileSync(outputPath, lines.join('\n'), 'utf8');
}

// ──────────────────────────────────────────────────────────────────
// processData – Core logic: filter, report, and write cleaned data
//   @param {string} filePath   – source CSV file path
//   @param {string} outputPath – destination CSV file path
// ──────────────────────────────────────────────────────────────────
function processData(filePath, outputPath) {
    // Step 1: Read and parse the CSV
    const records = readCSV(filePath);

    if (records.length === 0) {
        console.error('[ERROR] No valid records found in the file.');
        process.exit(1);
    }

    console.log(`\nTotal records loaded: ${records.length}`);

    // Step 2: Separate clean records from records with empty fields
    const cleanRecords   = [];
    const removedRecords = [];

    for (const r of records) {
        const emptyFields = getEmptyFields(r);
        if (emptyFields.length > 0) {
            removedRecords.push({ record: r, emptyFields }); // mark for removal
        } else {
            cleanRecords.push(r); // keep this record
        }
    }

    // Step 3: Display removed rows
    console.log();
    console.log('─────────────────────────────────────────────────────────');
    console.log(' ROWS REMOVED (contain at least one empty field):');
    console.log('─────────────────────────────────────────────────────────');

    if (removedRecords.length === 0) {
        console.log(' ✔ No rows with empty fields found.');
    } else {
        for (const item of removedRecords) {
            const name = item.record.candidate || '(no name)';
            console.log(` • ${name.padEnd(30)} | Empty: ${item.emptyFields.join(', ')}`);
        }
    }

    // Step 4: Print summary
    console.log();
    console.log('─────────────────────────────────────────────────────────');
    console.log(' SUMMARY');
    console.log('─────────────────────────────────────────────────────────');
    console.log(` Total records   : ${records.length}`);
    console.log(` Removed (empty) : ${removedRecords.length}`);
    console.log(` Clean records   : ${cleanRecords.length}`);

    // Step 5: Write cleaned data to output file
    writeCleanCSV(cleanRecords, outputPath);
    console.log();
    console.log(`✔ Cleaned dataset saved to: ${outputPath}`);
}

// ──────────────────────────────────────────────────────────────────
// MAIN – Entry point: prompts user for input and output paths
// ──────────────────────────────────────────────────────────────────
function main() {
    console.log('╔══════════════════════════════════════════════════════╗');
    console.log('║       MP18 – Remove Rows with Empty Fields           ║');
    console.log('╚══════════════════════════════════════════════════════╝');
    console.log();

    // Create readline interface for interactive console input
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

    // Step 1: Ask for source CSV path
    rl.question('Enter the path to the CSV dataset file: ', (filePath) => {
        filePath = filePath.trim();

        // Validate source file exists
        if (!fs.existsSync(filePath)) {
            console.error(`[ERROR] File not found: ${filePath}`);
            rl.close();
            process.exit(1);
        }

        // Step 2: Ask for output file path
        rl.question('Enter output file path for cleaned CSV (e.g., cleaned_data.csv): ', (outputPath) => {
            rl.close();
            outputPath = outputPath.trim();

            try {
                processData(filePath, outputPath);
            } catch (err) {
                console.error('[ERROR] Failed to process file:', err.message);
                process.exit(1);
            }
        });
    });
}

// Execute program
main();

File: MIDTERM-LAB-WORK-3/MP-18/removeemptyrows.java
/**
 * MP18 – Remove Rows with Empty Fields
 *
 * This program reads the CSV exam dataset, identifies and removes any rows
 * that contain one or more empty/blank fields, then saves the cleaned data
 * to a new CSV file. A summary of removed vs retained rows is displayed.
 *
 * Dataset: Exam results from University of Perpetual Help System - Molino
 * Columns: Candidate, Type, Exam, Language, Exam Date, Score, Result, Time Used
 */

import java.io.*;
import java.util.*;

public class MP18_RemoveEmptyRows {

    // ---------------------------------------------------------------
    // Inner class to hold a parsed exam record plus its raw CSV line
    // ---------------------------------------------------------------
    static class ExamRecord {
        String candidate;   // Candidate's name
        String type;        // Student / Faculty / NTE
        String exam;        // Exam subject/title
        String language;    // Language used
        String examDate;    // Date of exam
        String score;       // Numeric score
        String result;      // PASS or FAIL
        String timeUsed;    // Time taken
        String rawLine;     // Original CSV line (used when writing output)

        ExamRecord(String candidate, String type, String exam,
                   String language, String examDate,
                   String score, String result,
                   String timeUsed, String rawLine) {
            this.candidate = candidate.trim();
            this.type      = type.trim();
            this.exam      = exam.trim();
            this.language  = language.trim();
            this.examDate  = examDate.trim();
            this.score     = score.trim();
            this.result    = result.trim();
            this.timeUsed  = timeUsed.trim();
            this.rawLine   = rawLine;
        }

        // hasEmptyField – returns true if any essential field is blank
        boolean hasEmptyField() {
            return candidate.isEmpty() || type.isEmpty()     ||
                   exam.isEmpty()      || language.isEmpty() ||
                   examDate.isEmpty()  || score.isEmpty()    ||
                   result.isEmpty()    || timeUsed.isEmpty();
        }

        // emptyFieldsList – returns names of empty fields for reporting
        List<String> emptyFieldsList() {
            List<String> empty = new ArrayList<>();
            if (candidate.isEmpty()) empty.add("Candidate");
            if (type.isEmpty())      empty.add("Type");
            if (exam.isEmpty())      empty.add("Exam");
            if (language.isEmpty())  empty.add("Language");
            if (examDate.isEmpty())  empty.add("Exam Date");
            if (score.isEmpty())     empty.add("Score");
            if (result.isEmpty())    empty.add("Result");
            if (timeUsed.isEmpty())  empty.add("Time Used");
            return empty;
        }
    }

    // ---------------------------------------------------------------
    // MAIN – Program entry point
    // ---------------------------------------------------------------
    public static void main(String[] args) {

        System.out.println("╔══════════════════════════════════════════════════════╗");
        System.out.println("║       MP18 – Remove Rows with Empty Fields           ║");
        System.out.println("╚══════════════════════════════════════════════════════╝");
        System.out.println();

        // Step 1: Prompt user for the CSV file path
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the path to the CSV dataset file: ");
        String filePath = scanner.nextLine().trim();

        // Step 2: Read and parse the CSV
        List<ExamRecord> records = readCSV(filePath);

        if (records.isEmpty()) {
            System.out.println("[ERROR] No valid records found. Please check the file.");
            scanner.close();
            return;
        }

        System.out.println("\nTotal records loaded: " + records.size());

        // Step 3: Separate complete rows from rows with empty fields
        List<ExamRecord> cleanRecords  = new ArrayList<>();
        List<ExamRecord> removedRecords = new ArrayList<>();

        for (ExamRecord r : records) {
            if (r.hasEmptyField()) {
                removedRecords.add(r);
            } else {
                cleanRecords.add(r);
            }
        }

        // Step 4: Display removed rows
        System.out.println();
        System.out.println("─────────────────────────────────────────────────────────");
        System.out.println(" ROWS REMOVED (contain at least one empty field):");
        System.out.println("─────────────────────────────────────────────────────────");

        if (removedRecords.isEmpty()) {
            System.out.println(" ✔ No rows with empty fields found.");
        } else {
            for (ExamRecord r : removedRecords) {
                System.out.printf(" • %-30s | Empty: %s%n",
                    r.candidate.isEmpty() ? "(no name)" : r.candidate,
                    String.join(", ", r.emptyFieldsList()));
            }
        }

        // Step 5: Display summary
        System.out.println();
        System.out.println("─────────────────────────────────────────────────────────");
        System.out.println(" SUMMARY");
        System.out.println("─────────────────────────────────────────────────────────");
        System.out.printf(" Total records   : %d%n", records.size());
        System.out.printf(" Removed (empty) : %d%n", removedRecords.size());
        System.out.printf(" Clean records   : %d%n", cleanRecords.size());

        // Step 6: Write cleaned records to output file
        System.out.print("\nEnter output file path for cleaned CSV (e.g., cleaned_data.csv): ");
        String outputPath = scanner.nextLine().trim();

        writeCleanCSV(cleanRecords, outputPath);

        System.out.println();
        System.out.println("✔ Cleaned dataset saved to: " + outputPath);

        scanner.close();
    }

    // ---------------------------------------------------------------
    // readCSV – Reads and parses the CSV file into ExamRecord list
    //   @param filePath – path to input CSV file
    //   @return         – list of ExamRecord objects
    // ---------------------------------------------------------------
    static List<ExamRecord> readCSV(String filePath) {
        List<ExamRecord> records = new ArrayList<>();

        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {

            String line;
            boolean headerFound = false; // flag: passed the data-header row

            while ((line = br.readLine()) != null) {

                if (line.trim().isEmpty()) continue;

                // Detect the actual column-header row
                if (!headerFound) {
                    if (line.contains("Candidate")) {
                        headerFound = true;
                    }
                    continue;
                }

                // Parse the data row
                String[] tokens = parseCSVLine(line);
                if (tokens.length < 9) continue;

                String candidate = tokens[0];
                String type      = tokens[1];
                String exam      = tokens[3];
                String language  = tokens[4];
                String examDate  = tokens[5];
                String score     = tokens[6];
                String result    = tokens[7];
                String timeUsed  = (tokens.length > 8) ? tokens[8] : "";

                // Create record even if some fields are empty (we need to detect them)
                records.add(new ExamRecord(
                    candidate, type, exam, language,
                    examDate, score, result, timeUsed, line
                ));
            }

        } catch (FileNotFoundException e) {
            System.out.println("[ERROR] File not found: " + filePath);
        } catch (IOException e) {
            System.out.println("[ERROR] Could not read file: " + e.getMessage());
        }

        return records;
    }

    // ---------------------------------------------------------------
    // writeCleanCSV – Writes only clean records to a new CSV file
    //   @param records    – list of records with no empty fields
    //   @param outputPath – destination file path
    // ---------------------------------------------------------------
    static void writeCleanCSV(List<ExamRecord> records, String outputPath) {
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(outputPath))) {

            // Write CSV header row
            bw.write("Candidate,Type,Exam,Language,Exam Date,Score,Result,Time Used");
            bw.newLine();

            // Write each clean record as a CSV row
            for (ExamRecord r : records) {
                // Wrap fields that may contain commas in double quotes
                bw.write(String.format("\"%s\",\"%s\",\"%s\",\"%s\",\"%s\",\"%s\",\"%s\",\"%s\"",
                    r.candidate, r.type, r.exam, r.language,
                    r.examDate, r.score, r.result, r.timeUsed));
                bw.newLine();
            }

        } catch (IOException e) {
            System.out.println("[ERROR] Could not write output file: " + e.getMessage());
        }
    }

    // ---------------------------------------------------------------
    // parseCSVLine – Splits CSV line respecting quoted fields
    //   @param line – raw CSV line
    //   @return     – array of field strings
    // ---------------------------------------------------------------
    static String[] parseCSVLine(String line) {
        List<String> tokens = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean inQuotes = false;

        for (char c : line.toCharArray()) {
            if (c == '"') {
                inQuotes = !inQuotes;
            } else if (c == ',' && !inQuotes) {
                tokens.add(current.toString().trim());
                current.setLength(0);
            } else {
                current.append(c);
            }
        }
        tokens.add(current.toString().trim());

        return tokens.toArray(new String[0]);
    }
}